src/concurrency/genmc/intercept.rs

```diff
@@ -1,16 +1,136 @@
+use rustc_middle::throw_unsup_format;
 use tracing::debug;
 
 use crate::concurrency::thread::EvalContextExt as _;
 use crate::{
-    BlockReason, InterpResult, MachineCallback, MiriInterpCx, OpTy, UnblockKind, VisitProvenance,
-    VisitWith, callback, interp_ok,
+    BlockReason, InterpResult, MachineCallback, MiriInterpCx, OpTy, Scalar, UnblockKind,
+    VisitProvenance, VisitWith, callback, interp_ok, throw_ub_format,
 };
 
 // Handling of code intercepted by Miri in GenMC mode, such as assume statement or `std::sync::Mutex`.
 
-/// Other functionality not directly related to event handling
 impl<'tcx> EvalContextExt<'tcx> for crate::MiriInterpCx<'tcx> {}
 pub trait EvalContextExt<'tcx>: crate::MiriInterpCxExt<'tcx> {
+    /// Given a `ty::Instance<'tcx>`, do any required special handling.
+    /// Returns true if this `instance` should be skipped (i.e., no MIR should be executed for it).
+    fn check_genmc_intercept_function(
+        &mut self,
+        instance: rustc_middle::ty::Instance<'tcx>,
+        args: &[rustc_const_eval::interpret::FnArg<'tcx, crate::Provenance>],
+        dest: &crate::PlaceTy<'tcx>,
+    ) -> InterpResult<'tcx, bool> {
+        let this = self.eval_context_mut();
+        let genmc_ctx = this
+            .machine
+            .data_race
+            .as_genmc_ref()
+            .expect("This function should only be called in GenMC mode.");
+
+        let get_mutex_call_infos = || {
+            assert_eq!(args.len(), 1);
+            let arg = this.copy_fn_arg(&args[0]);
+            let addr = this.read_target_usize(&arg)?;
+            // FIXME(genmc): assert that we have at least 1 byte.
+            // FIXME(genmc): maybe use actual size of mutex here?.
+
+            let thread_infos = genmc_ctx.exec_state.thread_id_manager.borrow();
+            let curr_thread = this.machine.threads.active_thread();
+            let genmc_curr_thread = thread_infos.get_genmc_tid(curr_thread);
```

Comment:
This looks like you may want a little helper `active_thread_genmc_tid` or so.

And then the closure isn't worth its weight any more. It's anyway usually easier to follow when each shim starts by parsing its arguments, thus making it clear what the signature is. You probably want a little helper for that, something similar to `check_shim_sig_lenient`.